In [2]:
import pandas as pd
import numpy as np
import optuna
from plotly.io import show
import optuna.visualization as ov
from sklearn.preprocessing import StandardScaler
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.inspection import DecisionBoundaryDisplay
import re

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
#importazione dei dataset
df = pd.read_csv("train_titanic.csv")

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [5]:
df = df.drop(columns=['Cabin', 'Name'])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(3)
memory usage: 69.7+ KB


In [6]:
def extract_numeric(x):
    if pd.isna(x):
        return np.nan
    s = str(x)
    m = re.findall(r"[-+]?\d*\.?\d+", s)
    if not m:
        return np.nan
    return float(m[0])

numeric_cols_raw = [
    "Ticket",
]

for col in numeric_cols_raw:
    if col in df.columns:
        df[col] = df[col].apply(extract_numeric)
        
df["Age"] = df["Age"].fillna(df["Age"].median())

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       887 non-null    float64
 8   Fare         891 non-null    float64
 9   Embarked     889 non-null    object 
dtypes: float64(3), int64(5), object(2)
memory usage: 69.7+ KB


In [7]:
#rimozione outlier
numeric_cols = df.select_dtypes(include=["float", "int"]).columns.tolist()

for col in numeric_cols:
    Q1 = df[col].quantile(0.2)
    Q3 = df[col].quantile(0.8)
    IQR = Q3 - Q1
    lower, upper = Q1 - 1.5 * IQR, Q3 + 1.5 * IQR

    df[col] = np.where(df[col].between(lower, upper), df[col], np.nan)

df = df.dropna()

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 741 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  741 non-null    float64
 1   Survived     741 non-null    float64
 2   Pclass       741 non-null    float64
 3   Sex          741 non-null    object 
 4   Age          741 non-null    float64
 5   SibSp        741 non-null    float64
 6   Parch        741 non-null    float64
 7   Ticket       741 non-null    float64
 8   Fare         741 non-null    float64
 9   Embarked     741 non-null    object 
dtypes: float64(8), object(2)
memory usage: 63.7+ KB


In [8]:
df = pd.get_dummies(df, drop_first=True)

In [9]:
def objective(trial):
    params = {
        'objective': 'binary:logistic',
        'n_estimators': trial.suggest_int('n_estimators', 100, 1000), # Intero tra 100 e 1000
        'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.3, log=True), # Scala logaritmica
        'max_depth': trial.suggest_int('max_depth', 3, 10),
        'subsample': trial.suggest_float('subsample', 0.5, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.5, 1.0), # % feature per albero
        'reg_alpha': trial.suggest_float('reg_alpha', 1e-8, 10.0, log=True), # L1 Reg (Lasso)
        'reg_lambda': trial.suggest_float('reg_lambda', 1e-8, 10.0, log=True), # L2 Reg (Ridge)
        'n_jobs': 3,
        'random_state': 42
    }
    
    model = xgb.XGBClassifier(**params)
    
    # Usiamo 5-Fold per essere sicuri che il risultato non sia fortuna
    cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
    scores = cross_val_score(model, X_train, y_train, cv=cv, scoring='accuracy')
    
    if trial.should_prune():
            raise optuna.TrialPruned()
    
    # Restituiamo la media dell'accuratezza
    return scores.mean()

In [10]:
X = df.drop('Survived', axis=1)
y = df["Survived"].astype(int)

In [11]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [12]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

**OPTUNA**

In [16]:
sampler = optuna.samplers.TPESampler(seed=10)
study = optuna.create_study(direction="maximize", sampler = sampler,
    pruner=optuna.pruners.MedianPruner(
        n_startup_trials=5, n_warmup_steps=30, interval_steps=10
    ),
)
study.optimize(objective, n_trials=50, show_progress_bar=True)

print("\n--- Risultati Ottimizzazione ---")
print(f"Miglior Trial (Tentativo #{study.best_trial.number})")
print(f"Accuratezza Migliore (CV): {study.best_value:.4f}")
print("Migliori Iperparametri:")
for key, value in study.best_params.items():
    print(f"  {key}: {value}")

best_params = study.best_params
final_model = xgb.XGBClassifier(**best_params, n_jobs=3, random_state=42)
final_model.fit(X_train, y_train)
final_acc = final_model.score(X_test, y_test)

print(f"\nAccuratezza Finale sul Test Set: {final_acc:.4f}")

[I 2025-12-05 12:56:10,484] A new study created in memory with name: no-name-dd68316e-c214-47fc-87dd-d28e5d290bf9
Best trial: 0. Best value: 0.810825:   2%|▏         | 1/50 [00:02<02:09,  2.64s/it]

[I 2025-12-05 12:56:13,129] Trial 0 finished with value: 0.8108246688505911 and parameters: {'n_estimators': 794, 'learning_rate': 0.00112565445587888, 'max_depth': 8, 'subsample': 0.8744019412693059, 'colsample_bytree': 0.7492535061512953, 'reg_alpha': 1.0547992438188775e-06, 'reg_lambda': 6.061300044367956e-07}. Best is trial 0 with value: 0.8108246688505911.


Best trial: 0. Best value: 0.810825:   4%|▍         | 2/50 [00:03<01:21,  1.69s/it]

[I 2025-12-05 12:56:14,151] Trial 1 finished with value: 0.8056971941318901 and parameters: {'n_estimators': 785, 'learning_rate': 0.00262366298138214, 'max_depth': 3, 'subsample': 0.8426799091838986, 'colsample_bytree': 0.9766966730974682, 'reg_alpha': 1.08526150100961e-08, 'reg_lambda': 0.0004071274363191098}. Best is trial 0 with value: 0.8108246688505911.


Best trial: 0. Best value: 0.810825:   6%|▌         | 3/50 [00:05<01:27,  1.87s/it]

[I 2025-12-05 12:56:16,229] Trial 2 finished with value: 0.7938755163082182 and parameters: {'n_estimators': 832, 'learning_rate': 0.032907988677835835, 'max_depth': 8, 'subsample': 0.6459380340853166, 'colsample_bytree': 0.9588870612564717, 'reg_alpha': 0.026988705266625616, 'reg_lambda': 0.0007636587145416643}. Best is trial 0 with value: 0.8108246688505911.


Best trial: 3. Best value: 0.8142:   8%|▊         | 4/50 [00:06<01:05,  1.42s/it]  

[I 2025-12-05 12:56:16,955] Trial 3 finished with value: 0.8142002563737358 and parameters: {'n_estimators': 228, 'learning_rate': 0.008410277620965793, 'max_depth': 8, 'subsample': 0.720916587211498, 'colsample_bytree': 0.7170069966666468, 'reg_alpha': 0.003629968081041487, 'reg_lambda': 0.0004151874170225355}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  10%|█         | 5/50 [00:08<01:11,  1.59s/it]

[I 2025-12-05 12:56:18,857] Trial 4 finished with value: 0.7955704315624555 and parameters: {'n_estimators': 686, 'learning_rate': 0.030820974516764386, 'max_depth': 9, 'subsample': 0.7608235761968171, 'colsample_bytree': 0.9543244404043341, 'reg_alpha': 7.4666329224550155e-06, 'reg_lambda': 6.51829697963914e-08}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  12%|█▏        | 6/50 [00:09<01:04,  1.46s/it]

[I 2025-12-05 12:56:20,069] Trial 5 finished with value: 0.800726392251816 and parameters: {'n_estimators': 370, 'learning_rate': 0.001915812865819437, 'max_depth': 9, 'subsample': 0.5234481596946249, 'colsample_bytree': 0.8131435741556963, 'reg_alpha': 0.0008477648448701759, 'reg_lambda': 0.23636488716609783}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  14%|█▍        | 7/50 [00:10<00:50,  1.16s/it]

[I 2025-12-05 12:56:20,616] Trial 6 finished with value: 0.7955846745477853 and parameters: {'n_estimators': 279, 'learning_rate': 0.13259345778487028, 'max_depth': 5, 'subsample': 0.8773238457649286, 'colsample_bytree': 0.6479808534398394, 'reg_alpha': 0.902460736457843, 'reg_lambda': 8.503637539944912e-06}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  16%|█▌        | 8/50 [00:10<00:37,  1.11it/s]

[I 2025-12-05 12:56:20,945] Trial 7 finished with value: 0.8040592508189717 and parameters: {'n_estimators': 248, 'learning_rate': 0.009383017650421821, 'max_depth': 3, 'subsample': 0.9105528289184642, 'colsample_bytree': 0.5755760098212819, 'reg_alpha': 2.8643760792886104e-05, 'reg_lambda': 3.1502600820415925}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  18%|█▊        | 9/50 [00:13<00:58,  1.43s/it]

[I 2025-12-05 12:56:23,552] Trial 8 finished with value: 0.8006979062811566 and parameters: {'n_estimators': 989, 'learning_rate': 0.013499624615859645, 'max_depth': 9, 'subsample': 0.6256870671035297, 'colsample_bytree': 0.7986858241154422, 'reg_alpha': 1.335014070421493, 'reg_lambda': 0.0006471747121510108}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  20%|██        | 10/50 [00:14<00:54,  1.37s/it]

[I 2025-12-05 12:56:24,798] Trial 9 finished with value: 0.8024070645207235 and parameters: {'n_estimators': 631, 'learning_rate': 0.0012511393991761941, 'max_depth': 5, 'subsample': 0.5398065450779821, 'colsample_bytree': 0.6527299591714091, 'reg_alpha': 9.47270053228608e-06, 'reg_lambda': 0.09214518915553625}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  22%|██▏       | 11/50 [00:15<00:46,  1.19s/it]

[I 2025-12-05 12:56:25,581] Trial 10 finished with value: 0.7820538384845465 and parameters: {'n_estimators': 461, 'learning_rate': 0.2632546106931027, 'max_depth': 6, 'subsample': 0.7423309215793747, 'colsample_bytree': 0.5190212538232342, 'reg_alpha': 0.004965324688397551, 'reg_lambda': 8.055423019349575e-06}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  24%|██▍       | 12/50 [00:15<00:36,  1.03it/s]

[I 2025-12-05 12:56:26,051] Trial 11 finished with value: 0.807463324312776 and parameters: {'n_estimators': 147, 'learning_rate': 0.00455065796860278, 'max_depth': 7, 'subsample': 0.9658921357553798, 'colsample_bytree': 0.7276253587863623, 'reg_alpha': 1.9378747914410866e-07, 'reg_lambda': 1.417944344319233e-08}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  26%|██▌       | 13/50 [00:17<00:47,  1.28s/it]

[I 2025-12-05 12:56:28,037] Trial 12 finished with value: 0.8141717704030764 and parameters: {'n_estimators': 501, 'learning_rate': 0.005431665501072537, 'max_depth': 10, 'subsample': 0.7781402885174084, 'colsample_bytree': 0.8457528142671924, 'reg_alpha': 3.318614345546793e-07, 'reg_lambda': 1.794628119622844e-06}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 3. Best value: 0.8142:  28%|██▊       | 14/50 [00:19<00:53,  1.49s/it]

[I 2025-12-05 12:56:29,995] Trial 13 finished with value: 0.8107961828799317 and parameters: {'n_estimators': 496, 'learning_rate': 0.006366823257435648, 'max_depth': 10, 'subsample': 0.7598527191428301, 'colsample_bytree': 0.8684526041657525, 'reg_alpha': 0.07445973275148882, 'reg_lambda': 1.2656860213733482e-05}. Best is trial 3 with value: 0.8142002563737358.


Best trial: 14. Best value: 0.817547:  30%|███       | 15/50 [00:20<00:41,  1.19s/it]

[I 2025-12-05 12:56:30,507] Trial 14 finished with value: 0.8175473579262214 and parameters: {'n_estimators': 126, 'learning_rate': 0.026350251404998647, 'max_depth': 10, 'subsample': 0.6882430750004905, 'colsample_bytree': 0.8717752628663552, 'reg_alpha': 0.0012698599954344425, 'reg_lambda': 0.01225113959333315}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  32%|███▏      | 16/50 [00:20<00:31,  1.07it/s]

[I 2025-12-05 12:56:30,856] Trial 15 finished with value: 0.8056971941318901 and parameters: {'n_estimators': 117, 'learning_rate': 0.03819605608281092, 'max_depth': 7, 'subsample': 0.6681192701692128, 'colsample_bytree': 0.9002250092740622, 'reg_alpha': 0.0002455068711307922, 'reg_lambda': 0.013132963081130957}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  34%|███▍      | 17/50 [00:21<00:29,  1.12it/s]

[I 2025-12-05 12:56:31,644] Trial 16 finished with value: 0.7870958552912691 and parameters: {'n_estimators': 260, 'learning_rate': 0.07424055556271217, 'max_depth': 10, 'subsample': 0.6866785325799772, 'colsample_bytree': 0.698126824046296, 'reg_alpha': 0.0002853639432901908, 'reg_lambda': 0.00585494602580717}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  36%|███▌      | 18/50 [00:22<00:29,  1.10it/s]

[I 2025-12-05 12:56:32,588] Trial 17 finished with value: 0.8057684090585386 and parameters: {'n_estimators': 359, 'learning_rate': 0.018123098558590844, 'max_depth': 8, 'subsample': 0.5650614750974303, 'colsample_bytree': 0.7903411007454051, 'reg_alpha': 0.055724908538259885, 'reg_lambda': 6.468660190732053}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  38%|███▊      | 19/50 [00:22<00:22,  1.39it/s]

[I 2025-12-05 12:56:32,871] Trial 18 finished with value: 0.7888477424868253 and parameters: {'n_estimators': 190, 'learning_rate': 0.060689405259344714, 'max_depth': 9, 'subsample': 0.6047366375719402, 'colsample_bytree': 0.6702940384608902, 'reg_alpha': 8.091051783962001, 'reg_lambda': 5.976239615344023e-05}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  40%|████      | 20/50 [00:22<00:18,  1.65it/s]

[I 2025-12-05 12:56:33,210] Trial 19 finished with value: 0.8158809286426434 and parameters: {'n_estimators': 101, 'learning_rate': 0.01641148917532061, 'max_depth': 8, 'subsample': 0.7038468700609458, 'colsample_bytree': 0.8985862162647514, 'reg_alpha': 0.005085496007773075, 'reg_lambda': 0.03486038894662846}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  42%|████▏     | 21/50 [00:23<00:20,  1.41it/s]

[I 2025-12-05 12:56:34,163] Trial 20 finished with value: 0.8074063523714571 and parameters: {'n_estimators': 376, 'learning_rate': 0.018483523923183903, 'max_depth': 6, 'subsample': 0.7912763499396327, 'colsample_bytree': 0.9001284795361878, 'reg_alpha': 0.0024876117198248024, 'reg_lambda': 0.3411485658147558}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  44%|████▍     | 22/50 [00:24<00:17,  1.62it/s]

[I 2025-12-05 12:56:34,569] Trial 21 finished with value: 0.8107961828799317 and parameters: {'n_estimators': 112, 'learning_rate': 0.010747889906395908, 'max_depth': 8, 'subsample': 0.7084147702342456, 'colsample_bytree': 0.9139080407780706, 'reg_alpha': 0.007904623418559813, 'reg_lambda': 0.003939047291122082}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  46%|████▌     | 23/50 [00:24<00:16,  1.63it/s]

[I 2025-12-05 12:56:35,175] Trial 22 finished with value: 0.8125195841048283 and parameters: {'n_estimators': 193, 'learning_rate': 0.021864816239310088, 'max_depth': 7, 'subsample': 0.7131229451350575, 'colsample_bytree': 0.8497008000437571, 'reg_alpha': 7.404493772032188e-05, 'reg_lambda': 0.02870221954693563}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  48%|████▊     | 24/50 [00:25<00:20,  1.25it/s]

[I 2025-12-05 12:56:36,400] Trial 23 finished with value: 0.8091155106110242 and parameters: {'n_estimators': 292, 'learning_rate': 0.008267058733494282, 'max_depth': 10, 'subsample': 0.5874697158355595, 'colsample_bytree': 0.9965425750592364, 'reg_alpha': 0.23794732635291765, 'reg_lambda': 0.0025339721720661857}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  50%|█████     | 25/50 [00:26<00:19,  1.32it/s]

[I 2025-12-05 12:56:37,071] Trial 24 finished with value: 0.7989602620709302 and parameters: {'n_estimators': 204, 'learning_rate': 0.06831912202182391, 'max_depth': 9, 'subsample': 0.8268935812784051, 'colsample_bytree': 0.9291154897834182, 'reg_alpha': 0.0008673612093336129, 'reg_lambda': 6.790353970373853e-05}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  52%|█████▏    | 26/50 [00:26<00:15,  1.55it/s]

[I 2025-12-05 12:56:37,451] Trial 25 finished with value: 0.7669135450790485 and parameters: {'n_estimators': 122, 'learning_rate': 0.004441589064908324, 'max_depth': 8, 'subsample': 0.6582619664116264, 'colsample_bytree': 0.6049561988753354, 'reg_alpha': 0.012353805017072476, 'reg_lambda': 1.1621131878446689}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  54%|█████▍    | 27/50 [00:27<00:15,  1.46it/s]

[I 2025-12-05 12:56:38,226] Trial 26 finished with value: 0.8091439965816836 and parameters: {'n_estimators': 315, 'learning_rate': 0.003044851777694877, 'max_depth': 6, 'subsample': 0.72193531606941, 'colsample_bytree': 0.7666117052210742, 'reg_alpha': 0.0013360410311401467, 'reg_lambda': 0.04485335016263642}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  56%|█████▌    | 28/50 [00:28<00:14,  1.55it/s]

[I 2025-12-05 12:56:38,783] Trial 27 finished with value: 0.7939182452642075 and parameters: {'n_estimators': 199, 'learning_rate': 0.046453578407820356, 'max_depth': 7, 'subsample': 0.6675884606477618, 'colsample_bytree': 0.8221374892307416, 'reg_alpha': 0.00010535091579993895, 'reg_lambda': 0.0001458260216129609}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  58%|█████▊    | 29/50 [00:29<00:14,  1.50it/s]

[I 2025-12-05 12:56:39,502] Trial 28 finished with value: 0.8040734938043015 and parameters: {'n_estimators': 405, 'learning_rate': 0.02413421565717828, 'max_depth': 4, 'subsample': 0.7897363682324369, 'colsample_bytree': 0.8749164189650009, 'reg_alpha': 0.08549395627850412, 'reg_lambda': 0.002053455295182002}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  60%|██████    | 30/50 [00:29<00:11,  1.72it/s]

[I 2025-12-05 12:56:39,878] Trial 29 finished with value: 0.7938755163082183 and parameters: {'n_estimators': 100, 'learning_rate': 0.12654341238608308, 'max_depth': 9, 'subsample': 0.8257749804017144, 'colsample_bytree': 0.733275331882606, 'reg_alpha': 0.3382964160946496, 'reg_lambda': 0.466658174272973}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  62%|██████▏   | 31/50 [00:31<00:18,  1.02it/s]

[I 2025-12-05 12:56:41,787] Trial 30 finished with value: 0.8006551773251674 and parameters: {'n_estimators': 584, 'learning_rate': 0.012336823641217944, 'max_depth': 8, 'subsample': 0.6345886445447301, 'colsample_bytree': 0.7678081416869643, 'reg_alpha': 0.014605522566264588, 'reg_lambda': 0.012406952560666179}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  64%|██████▍   | 32/50 [00:33<00:22,  1.23s/it]

[I 2025-12-05 12:56:43,587] Trial 31 finished with value: 0.8091012676256943 and parameters: {'n_estimators': 451, 'learning_rate': 0.006054120879795518, 'max_depth': 10, 'subsample': 0.7760544135502059, 'colsample_bytree': 0.8441539053792309, 'reg_alpha': 3.975978491279459e-07, 'reg_lambda': 1.0859625863052991e-06}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  66%|██████▌   | 33/50 [00:35<00:28,  1.65s/it]

[I 2025-12-05 12:56:46,241] Trial 32 finished with value: 0.8074490813274462 and parameters: {'n_estimators': 716, 'learning_rate': 0.006912375363833721, 'max_depth': 10, 'subsample': 0.6947597685298363, 'colsample_bytree': 0.8756035701575012, 'reg_alpha': 5.529409135370239e-08, 'reg_lambda': 2.908696862991742e-07}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  68%|██████▊   | 34/50 [00:37<00:29,  1.83s/it]

[I 2025-12-05 12:56:48,484] Trial 33 finished with value: 0.8141575274177468 and parameters: {'n_estimators': 559, 'learning_rate': 0.0034885252328915313, 'max_depth': 10, 'subsample': 0.7374059012921703, 'colsample_bytree': 0.9307908896085528, 'reg_alpha': 6.699349480670785e-06, 'reg_lambda': 9.535735002790559e-07}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  70%|███████   | 35/50 [00:40<00:31,  2.07s/it]

[I 2025-12-05 12:56:51,128] Trial 34 finished with value: 0.8056829511465603 and parameters: {'n_estimators': 848, 'learning_rate': 0.012178659674540989, 'max_depth': 9, 'subsample': 0.804951571570229, 'colsample_bytree': 0.8329050020607062, 'reg_alpha': 0.003065597652226918, 'reg_lambda': 0.00027626264228514905}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  72%|███████▏  | 36/50 [00:41<00:22,  1.64s/it]

[I 2025-12-05 12:56:51,752] Trial 35 finished with value: 0.7770118216778237 and parameters: {'n_estimators': 179, 'learning_rate': 0.0018115795978357473, 'max_depth': 8, 'subsample': 0.8552524809726247, 'colsample_bytree': 0.9601292472603146, 'reg_alpha': 1.4719219583863172e-08, 'reg_lambda': 0.0010833383974693297}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  74%|███████▍  | 37/50 [00:42<00:17,  1.37s/it]

[I 2025-12-05 12:56:52,503] Trial 36 finished with value: 0.8107676969092722 and parameters: {'n_estimators': 228, 'learning_rate': 0.027955564960707533, 'max_depth': 9, 'subsample': 0.7488491328356005, 'colsample_bytree': 0.7021761427474059, 'reg_alpha': 0.0006046057443063763, 'reg_lambda': 0.10344628480148081}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  76%|███████▌  | 38/50 [00:42<00:13,  1.12s/it]

[I 2025-12-05 12:56:53,048] Trial 37 finished with value: 0.814200256373736 and parameters: {'n_estimators': 155, 'learning_rate': 0.016048755963244007, 'max_depth': 8, 'subsample': 0.6845455551831829, 'colsample_bytree': 0.8046214921182524, 'reg_alpha': 3.6646344882568117e-06, 'reg_lambda': 2.862218960720676e-05}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  78%|███████▊  | 39/50 [00:43<00:11,  1.09s/it]

[I 2025-12-05 12:56:54,059] Trial 38 finished with value: 0.8091155106110242 and parameters: {'n_estimators': 312, 'learning_rate': 0.015331091740310688, 'max_depth': 7, 'subsample': 0.6173612940933682, 'colsample_bytree': 0.7861958466276774, 'reg_alpha': 3.2707185000402227e-06, 'reg_lambda': 2.7535088667562363e-05}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  80%|████████  | 40/50 [00:44<00:09,  1.09it/s]

[I 2025-12-05 12:56:54,586] Trial 39 finished with value: 0.7955561885771258 and parameters: {'n_estimators': 158, 'learning_rate': 0.03581264596320949, 'max_depth': 8, 'subsample': 0.685437272902963, 'colsample_bytree': 0.749841709858265, 'reg_alpha': 3.5902098149567076e-05, 'reg_lambda': 0.0005457288580645337}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  82%|████████▏ | 41/50 [00:44<00:07,  1.24it/s]

[I 2025-12-05 12:56:55,134] Trial 40 finished with value: 0.8141575274177468 and parameters: {'n_estimators': 243, 'learning_rate': 0.009636178343942196, 'max_depth': 5, 'subsample': 0.5769313239468027, 'colsample_bytree': 0.7089367762951769, 'reg_alpha': 1.7404579646626454e-06, 'reg_lambda': 0.00013288269129287068}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  84%|████████▍ | 42/50 [00:45<00:06,  1.32it/s]

[I 2025-12-05 12:56:55,776] Trial 41 finished with value: 0.8158951716279732 and parameters: {'n_estimators': 152, 'learning_rate': 0.015805901826561004, 'max_depth': 9, 'subsample': 0.6457789717449763, 'colsample_bytree': 0.8097872233723078, 'reg_alpha': 4.5013004400571797e-07, 'reg_lambda': 2.6174978020719487e-07}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 14. Best value: 0.817547:  86%|████████▌ | 43/50 [00:45<00:04,  1.42it/s]

[I 2025-12-05 12:56:56,348] Trial 42 finished with value: 0.8125195841048285 and parameters: {'n_estimators': 145, 'learning_rate': 0.02047318401218673, 'max_depth': 9, 'subsample': 0.5047989115859096, 'colsample_bytree': 0.8105490325015513, 'reg_alpha': 9.15927665873663e-07, 'reg_lambda': 1.52832854675794e-07}. Best is trial 14 with value: 0.8175473579262214.


Best trial: 43. Best value: 0.819285:  88%|████████▊ | 44/50 [00:46<00:03,  1.55it/s]

[I 2025-12-05 12:56:56,864] Trial 43 finished with value: 0.8192850021364478 and parameters: {'n_estimators': 158, 'learning_rate': 0.015920226629661045, 'max_depth': 8, 'subsample': 0.6506614580881595, 'colsample_bytree': 0.7747822351341761, 'reg_alpha': 0.029165365785368588, 'reg_lambda': 1.799301891794392e-08}. Best is trial 43 with value: 0.8192850021364478.


Best trial: 44. Best value: 0.822632:  90%|█████████ | 45/50 [00:46<00:03,  1.63it/s]

[I 2025-12-05 12:56:57,406] Trial 44 finished with value: 0.8226321036889331 and parameters: {'n_estimators': 164, 'learning_rate': 0.015378390533926147, 'max_depth': 8, 'subsample': 0.6459385761190881, 'colsample_bytree': 0.7621269752553935, 'reg_alpha': 6.86748897958525e-08, 'reg_lambda': 2.1786988060483554e-08}. Best is trial 44 with value: 0.8226321036889331.


Best trial: 44. Best value: 0.822632:  92%|█████████▏| 46/50 [00:47<00:02,  1.50it/s]

[I 2025-12-05 12:56:58,187] Trial 45 finished with value: 0.8023500925794046 and parameters: {'n_estimators': 233, 'learning_rate': 0.02695743060932004, 'max_depth': 9, 'subsample': 0.6367402351022912, 'colsample_bytree': 0.7729952602926691, 'reg_alpha': 5.887414952324697e-08, 'reg_lambda': 1.3709655807207668e-08}. Best is trial 44 with value: 0.8226321036889331.


Best trial: 44. Best value: 0.822632:  94%|█████████▍| 47/50 [00:48<00:02,  1.27it/s]

[I 2025-12-05 12:56:59,266] Trial 46 finished with value: 0.7921948440393106 and parameters: {'n_estimators': 341, 'learning_rate': 0.03779529922050812, 'max_depth': 9, 'subsample': 0.6510563039576213, 'colsample_bytree': 0.863224665039376, 'reg_alpha': 1.051479297715986e-08, 'reg_lambda': 4.962795476190598e-08}. Best is trial 44 with value: 0.8226321036889331.


Best trial: 44. Best value: 0.822632:  96%|█████████▌| 48/50 [00:49<00:01,  1.21it/s]

[I 2025-12-05 12:57:00,179] Trial 47 finished with value: 0.8057399230878792 and parameters: {'n_estimators': 277, 'learning_rate': 0.049554857272946895, 'max_depth': 8, 'subsample': 0.6038254545065727, 'colsample_bytree': 0.8905824861672244, 'reg_alpha': 0.030306797583127643, 'reg_lambda': 5.669934907283949e-08}. Best is trial 44 with value: 0.8226321036889331.


Best trial: 44. Best value: 0.822632:  98%|█████████▊| 49/50 [00:50<00:00,  1.41it/s]

[I 2025-12-05 12:57:00,620] Trial 48 finished with value: 0.8192565161657883 and parameters: {'n_estimators': 147, 'learning_rate': 0.01405297404705908, 'max_depth': 7, 'subsample': 0.6479514962382967, 'colsample_bytree': 0.7490359022770384, 'reg_alpha': 4.179908061802853e-08, 'reg_lambda': 3.028950986554105e-08}. Best is trial 44 with value: 0.8226321036889331.


Best trial: 44. Best value: 0.822632: 100%|██████████| 50/50 [00:50<00:00,  1.01s/it]

[I 2025-12-05 12:57:01,088] Trial 49 finished with value: 0.8074490813274462 and parameters: {'n_estimators': 157, 'learning_rate': 0.009125766855110273, 'max_depth': 7, 'subsample': 0.5531558614083414, 'colsample_bytree': 0.7218307983749165, 'reg_alpha': 7.877410602857414e-08, 'reg_lambda': 3.7729241933368674e-08}. Best is trial 44 with value: 0.8226321036889331.

--- Risultati Ottimizzazione ---
Miglior Trial (Tentativo #44)
Accuratezza Migliore (CV): 0.8226
Migliori Iperparametri:
  n_estimators: 164
  learning_rate: 0.015378390533926147
  max_depth: 8
  subsample: 0.6459385761190881
  colsample_bytree: 0.7621269752553935
  reg_alpha: 6.86748897958525e-08
  reg_lambda: 2.1786988060483554e-08

Accuratezza Finale sul Test Set: 0.8523


Senza pruner: Accuratezza Finale sul Test Set: 0.8456

Con pruner: --- Risultati Ottimizzazione ---
- Miglior Trial (Tentativo #26)
- Accuratezza Migliore (CV): 0.8159
- Migliori Iperparametri:
  - n_estimators: 380
  - learning_rate: 0.005295232368540756
  - max_depth: 7
  - subsample: 0.6891247712682177
  - colsample_bytree: 0.6752990656146991
  - reg_alpha: 0.06532885224059053
  - reg_lambda: 0.00013423432607369992

- Accuratezza Finale sul Test Set: 0.8591

Outlier 0.2, 0.8: --- Risultati Ottimizzazione ---
- Miglior Trial (Tentativo #35)
- Accuratezza Migliore (CV): 0.8192
- Migliori Iperparametri:
  - n_estimators: 821
  - learning_rate: 0.0040452630812914565
  - max_depth: 6
  - subsample: 0.8144802291812205
  - colsample_bytree: 0.8504236185894885
  - reg_alpha: 0.0020024465144113522
  - reg_lambda: 0.013159768501168314

- Accuratezza Finale sul Test Set: 0.8523

**GRAFICO**

In [22]:
ov.plot_optimization_history(study).show()
ov.plot_param_importances(study).show()
ov.plot_parallel_coordinate(study).show()
ov.plot_contour(study).show()
ov.plot_intermediate_values(study).show()

[W 2025-12-05 13:01:28,723] You need to set up the pruning feature to utilize `plot_intermediate_values()`
